backend/db/refresh_dataset_group_tables.py

```diff
@@ -0,0 +1,127 @@
+        last_updated_us: str = check_db_status_var(config['last_updated_termhub_var'])
+        last_updated_them: str = get_last_update_of_dataset(config['last_updated_enclave_representative_table'])
+        if last_updated_us and last_updated_them and dp.parse(last_updated_us) > dp.parse(last_updated_them):
+            print('Tables already up to date. Nothing to be done.')
+            continue
+
+        # Download datasets
+        if not skip_downloads:
+            print('Downloading datasets')
+            download_datasets(single_group=group_name)
+
+        # Load data
+        if not download_only:
+            print('Loading downloaded datasets into DB tables')
+            con = get_db_connection(schema=schema)
+            # noinspection PyBroadException
+            try:
+                for table in config['tables']:  # tables themselves (non-derived)
+                    print(f' - loading {table}...')
+                    t0 = datetime.now()
+                    load_csv(
+                        con, table, replace_rule='do not replace', schema=schema, optional_suffix='_new',
+                        path_override=Path(alternate_dataset_dir) / f'{table}.csv' if alternate_dataset_dir else None,
+                        is_test_table=bool(alternate_dataset_dir))
+                    run_sql(con, f'ALTER TABLE IF EXISTS {schema}.{table} RENAME TO {table}_old;')
+                    run_sql(con, f'ALTER TABLE {schema}.{table}_new RENAME TO {table};')
+                    t1 = datetime.now()
+                    print(f'   done in {(t1 - t0).seconds} seconds')
+                    # todo: set variable for 'last updated' for each table (look at load())
+                    #  - consider: check if table already updated sooner than last_updated_them. if so, skip. and add a param to CLI for this
+
+                # TODO: PKs
+
+                print('Creating indexes')  # indexes
+                # TODO: Why didn't these indexes run?
+                statements: List[str] = get_ddl_statements(schema, ['indexes'], return_type='flat')
+                for statement in statements:
+                    try:
+                        run_sql(con, statement)
+                    except ProgrammingError as err:
+                        # ignore if non-existent table if running test (i.e. alternate_dataset_dir)
+                        if 'does not exist' not in str(err) and alternate_dataset_dir:
+                            raise err
+
+                print('Recreating derived tables')  # derived tables
+                refresh_derived_tables(con, config['tables'], schema)
+                print('Deleting old, temporarily backed up versions of tables')
+                for table in config['tables']:
+                    run_sql(con, f'DROP TABLE IF EXISTS {schema}.{table}_old;')
+            except Exception as err:
+                print('Error occurred during table refresh. Resetting tables to pre-refresh state; restoring backups.',
```

Comment:
## Resetting tables when error occurs
@Sigfried I think it would be nice if I had the time to set this up so that this exception handling occurred for all types of refreshes.

I've been wanting to do this for a while. Basically it is hella annoying, when I'm troubleshooting and in development, naturally errors occur, then I have to manually go and delete all the `_new` tables, and rename all the `_old` tables back to their original states. Finally automated that here.

Reply:
I have a new implementation, so closing this comment.